File: plugins/OrchidPosts/src/main/java/com/eden/orchid/posts/menu/CategoriesMenuType.java
package com.eden.orchid.posts.menu;

import com.eden.common.util.EdenPair;
import com.eden.common.util.EdenUtils;
import com.eden.orchid.api.OrchidContext;
import com.eden.orchid.api.options.annotations.Option;
import com.eden.orchid.api.theme.menus.menuItem.OrchidMenuItem;
import com.eden.orchid.api.theme.menus.menuItem.OrchidMenuItemImpl;
import com.eden.orchid.api.theme.pages.OrchidPage;
import com.eden.orchid.posts.PostsModel;
import com.eden.orchid.posts.pages.PostArchivePage;
import com.eden.orchid.posts.pages.PostPage;

import javax.inject.Inject;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class CategoriesMenuType extends OrchidMenuItem {

    private PostsModel postsModel;

    @Option
    public String category;

    @Option
    public String title;

    @Inject
    public CategoriesMenuType(OrchidContext context, PostsModel postsModel) {
        super(context, "postCategories", 100);

        this.postsModel = postsModel;
    }

    @Override
    public List<OrchidMenuItemImpl> getMenuItems() {
        List<OrchidMenuItemImpl> items = new ArrayList<>();

        if(!EdenUtils.isEmpty(category) && postsModel.getCategories().containsKey(category)) {
            EdenPair<List<PostPage>, List<PostArchivePage>> categoryDef = postsModel.getCategories().get(category);
            if(!EdenUtils.isEmpty(categoryDef.second)) {
                OrchidMenuItemImpl menuItem = new OrchidMenuItemImpl(context, categoryDef.second.get(0));
                if (!EdenUtils.isEmpty(title)) {
                    menuItem.setTitle(title);
                }
                items.add(menuItem);
            }
        }
        else if(postsModel.getCategoryNames().size() > 1) {
            List<OrchidPage> pages = new ArrayList<>();
            for(Map.Entry<String, EdenPair<List<PostPage>, List<PostArchivePage>>> tag : postsModel.getCategories().entrySet()) {
                if(!EdenUtils.isEmpty(tag.getValue().second)) {
                    pages.add(tag.getValue().second.get(0));
                }
            }

            OrchidMenuItemImpl menuItem = new OrchidMenuItemImpl(context, "Categories", pages);
            if(!EdenUtils.isEmpty(title)) {
                menuItem.setTitle(title);
            }
            items.add(menuItem);
        }
        else if(postsModel.getCategories().containsKey(null)) {
            EdenPair<List<PostPage>, List<PostArchivePage>> categoryDef = postsModel.getCategories().get(null);
            if(!EdenUtils.isEmpty(categoryDef.second)) {
                OrchidMenuItemImpl menuItem = new OrchidMenuItemImpl(context, categoryDef.second.get(0));
                if (!EdenUtils.isEmpty(title)) {
                    menuItem.setTitle(title);
                }
                items.add(menuItem);
            }
        }

        return items;
    }
}

File: plugins/OrchidPosts/src/main/kotlin/com/eden/orchid/posts/menu/CategoriesMenuType.kt
package com.eden.orchid.posts.menu

import com.eden.common.util.EdenUtils
import com.eden.orchid.api.OrchidContext
import com.eden.orchid.api.options.annotations.Option
import com.eden.orchid.api.theme.menus.menuItem.OrchidMenuItem
import com.eden.orchid.api.theme.menus.menuItem.OrchidMenuItemImpl
import com.eden.orchid.api.theme.pages.OrchidPage
import com.eden.orchid.posts.model.PostsModel
import java.util.*
import javax.inject.Inject

class CategoriesMenuType @Inject
constructor(context: OrchidContext, private val postsModel: PostsModel) : OrchidMenuItem(context, "postCategories", 100) {

    @Option
    lateinit var category: String

    @Option
    lateinit var title: String

    override fun getMenuItems(): List<OrchidMenuItemImpl> {
        val items = ArrayList<OrchidMenuItemImpl>()

        if (!EdenUtils.isEmpty(category) && postsModel.categories.containsKey(category)) {
            val categoryDef = postsModel.categories[category]
            if (!EdenUtils.isEmpty(categoryDef!!.second)) {
                val menuItem = OrchidMenuItemImpl(context, categoryDef.second[0])
                if (!EdenUtils.isEmpty(title)) {
                    menuItem.title = title
                }
                items.add(menuItem)
            }
        } else if (postsModel.categoryNames.size > 1) {
            val pages = ArrayList<OrchidPage>()
            for ((_, value) in postsModel.categories) {
                if (!EdenUtils.isEmpty(value.second)) {
                    pages.add(value.second[0])
                }
            }

            val menuItem = OrchidMenuItemImpl(context, "Categories", pages)
            if (!EdenUtils.isEmpty(title)) {
                menuItem.title = title
            }
            items.add(menuItem)
        } else if (postsModel.categories.containsKey(null)) {
            val categoryDef = postsModel.categories[null]
            if (!EdenUtils.isEmpty(categoryDef!!.second)) {
                val menuItem = OrchidMenuItemImpl(context, categoryDef.second[0])
                if (!EdenUtils.isEmpty(title)) {
                    menuItem.title = title
                }
                items.add(menuItem)
            }
        }

        return items
    }
}

